### PR TITLE
Fix problems with jdk detection

### PR DESCRIPTION
**Pull request by mwichmann**
The java tool common routine finds a jdk by doing a filesystem glob.  This had a problem on windows in the case a specific version is requested, because the format of name of the jdk directory has changed with JDK 9 - there is now a dash between jdk and the version string.  The glob which does not attempt to match a version was general enough not to trip on this, but with a version to match it would never match jdk-9 or higher. 
    
The test harness then asks the found `javac` what version it is, and the parsing of that did not work as expected once version numbers became double-digit, as the regex was for a single digit followed by a dot.  The outcome is for 11.0.2 we get back '11' instead of '11.0'.  Change the regex to match any number of digits followed by dot.
    
The `Repository/RMIC.py` change is to align with an earlier change to `Java/RMIC.py`, but this may not be needed after the change to regex just described.
    
Clean up some of the Java tool routines for consistency (there was no functional change outside of `JavaCommon.py`)
    
Docstrings added or updated in several places.

Per request, added a unit-test for the glob used to select a jdk version.

Signed-off-by: Mats Wichmann <mats@linux.com>

## Contributor Checklist:

* [X] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [X] I have updated `master/src/CHANGES.txt` directory (and read the `README.txt` in that directory)
* [X] I have updated the appropriate documentation

### CODE REVIEW

**Comment by mwichmann** [comment]
Added a test, but I'm not really sure this adds any long-term value.  If the globs change (in `testing/framework/TestSCons.py` and in `src/engine/SCons/Tool/JavaCommon.py`) this test would have to be updated manually.

**Comment by bdbaddog** [comment]
> Added a test, but I'm not really sure this adds any long-term value. If the globs change (in `testing/framework/TestSCons.py` and in `src/engine/SCons/Tool/JavaCommon.py`) this test would have to be updated manually.

In the past I've just added comments to all places that have to be changed in sync. Not ideal, but at least the info's near where a change would break things.

**Comment by bdbaddog** [comment]
I'd suggest a unit test for JavaCommon rather than a system test for testing the paths.
MUCH faster to run than creating a bunch of directories (especially on windows)..

**Comment by mwichmann** [comment]
Hmmm, I haven't done one of that kind of test.  Still need to create the test diretories to do the glob against, no (there's not any need for as much of a hierarchy as I made, though, really just need the actual versioned directories)

File: src/engine/SCons/Tool/JavaCommonTests.py [hunk 1]
```diff
@@ -608,6 +609,55 @@ def test_in_function_class_declaration(self):
...
+                        ['Program Files', 'Java'],
+                        ['Program Files', 'Java', 'jdk1.8.0_201'],
+                        ['Program Files', 'Java', 'jdk1.8.0_201', 'bin'],
+                        ['Program Files', 'Java', 'jdk-11.0.2'],
+                        ['Program Files', 'Java', 'jdk-11.0.2', 'bin'])
+
+            test.write(['Program Files', 'Java', 'jdk1.8.0_201', 'bin', 'javac.exe'], "echo Java 1.8")
+            test.write(['Program Files', 'Java', 'jdk-11.0.2', 'bin', 'javac.exe'], "echo Java 11.0")
+        else:
+            test.subdir(['jvm'],
```
Comment (bdbaddog) [inline src/engine/SCons/Tool/JavaCommonTests.py]: So should this be equivalent to abs path : /jvm/java/.... ? or */jva/java/...?

**Pull request merged**